tools/provision

```diff
@@ -24,10 +24,18 @@ the Python version this command is executed with."""
 
     venv_dir = os.path.join(base_dir, venv_name)
     if not os.path.isdir(venv_dir):
-        if subprocess.call(['virtualenv', '-p', python_interpreter, venv_dir]):
-            raise OSError("The command `virtualenv -p {} {}` failed. Virtualenv not created!"
-                          .format(python_interpreter, venv_dir))
+        try:
```

Comment:
Sounds great, I'll look into that. Thanks for the feedback!

Reply:
Unfortunately, I don't see how this can be applied here. This is a good way to check whether the code is running inside a virtual environment. What would be helpful in this case is a way to check whether the `virtualenv` package is installed (before using it to create an environment).

Reply:
Here is how it is checked in one of the tool in the Zulip webapp/server:
https://github.com/zulip/zulip/blob/master/tools/update-locked-requirements#L4-L6
If the `venv_dir` is known, a check along this line can be done.

Reply:
This is already being done one line above:
```
if not os.path.isdir(venv_dir):
    ...
```
But what if the environment doesn't exist? We have to run the following code to create it: 
```
subprocess.call(['virtualenv', '-p', python_interpreter, venv_dir])
```
To do it successfully we need to have the `virtualenv` package installed. On Ubuntu it can be done by running this command in the terminal:
```
sudo apt-get install python-virtualenv
```
But what if we make the `subprocess` call when the `virtualenv` package is not installed? The script will exit with an error and print the traceback I included above.

What I'm trying to do here is to replace this traceback with an error message that would tell people exactly what's going on: they need to install the package to run the script.

Reply:
Ic, wdyt about capturing if virtualenv is installed first before running it?
```
if not `which virtualenv`:
 print("... please install virtualenv")
```
Then, in the case when `virtualenv` has already been installed, leave the traceback as it is, since the exception message override you did masks away the detail of the problem (e.g. could be due to permission error).

Reply:
When the case is due to permission error, rerunning `virtualenv -p` won't solve the issue.

Reply:
I agree with the idea of checking if `virtualenv` is installed as the first step, makes the logic easier. Instead of `which virtualenv`, I suggest checking the exit code of `command -v virtualenv`, since its a shell built-in and part of posix.

Reply:
@rht @derAnfaenger Thanks for the suggestions! I agree that checking if the `virtualenv` package is installed before using it makes more sense than a try-except block. I'll make use of the `command -v virtualenv` too.

Reply:
Hm, the `subprocess` call fails when I replace `which virtualenv` with `command -v virtualenv`:
```
Traceback (most recent call last):
  File "./tools/provision", line 76, in <module>
    main()
  File "./tools/provision", line 26, in main
    if subprocess.call(['command', '-v', 'virtualenv']):
  File "/usr/lib/python2.7/subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1327, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```
Should I keep using `which virtualenv` or wrap it in a try-except block, perhaps?
```
try:
    subprocess.call(['command', '-v', 'virtualenv'])
except OSError:
    print("{red}Please install the virtualenv package and try again.{end_format}"
                  .format(red='\033[91m', end_format='\033[0m'))
    sys.exit(1)
```

Reply:
This discussion had a lot of suggestions that didn't actually work.  @rht next time, I'd encourage you to read the code more closely before commenting; e.g. https://stackoverflow.com/questions/1871549/python-determine-if-running-inside-virtualenv/1883251#1883251 was totally irrelevant to the problem at hand :)

Reply:
(my mistake :( )